## K-Means Model and Analysis

This notebook applies K-Means clustering to air quality data to identify groups of observations with similar characteristics, such as temperature, wind speed, precipitation, and fire-related features. 

The goal is to explore whether the data naturally forms distinct subgroups that might correspond to different environmental conditions or pollution patterns.

In [1]:
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline
from sklearn.cluster import KMeans
from sklearn.metrics import silhouette_score, mean_squared_error, r2_score
from sklearn.decomposition import PCA
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.linear_model import Ridge, Lasso
import plotly.express as px
import matplotlib.pyplot as plt
import os

In [7]:
script_dir = os.getcwd()
df = pd.read_csv(f'{script_dir}/air_quality_weather_fires.csv')

# Recoding weather codes into broader/more identifiable categories
df['weather_code'] = df['weather_code'].replace(['1', 'Clear sky', 'Mainly clear'], 'clear')
df = df[df.weather_code != '2']
df['weather_code'] = df['weather_code'].replace(
    ['3', 'Overcast', 'Partly cloudy'], 'cloudy')
df['weather_code'] = df['weather_code'].replace(
    ['51', '53', '55', '61', '63', '65', 
     'Dense drizzle', 'Heavy rain', 'Light drizzle', 'Moderate drizzle', 'Moderate rain', 'Slight rain'], 'rainy')
df['weather_code'] = df['weather_code'].replace(
    ['71', '73', '75', 'Heavy snow fall', 'Moderate snow fall', 'Slight snow fall'], 'snowy')

# Defining numeric variables for clustering and prediction
X_cols = ['latitude', 'longitude', 'temperature_2m_mean', 
          'wind_speed_10m_mean', 'precipitation_sum', 
          'fires_within_50km', 'fires_within_100km', 
          'distance_to_fire_km', 'fire_brightness']

# Converting columns to numeric and dropping rows with missing values
df[X_cols] = df[X_cols].apply(pd.to_numeric, errors='coerce')
df.dropna(subset=X_cols + ['PM25'], inplace=True)

X_data = df[X_cols]
y_data = df['PM25'].values

In [8]:
# K-means clustering pipeline
pipe = Pipeline([
    ('scaler', StandardScaler()),
    ('kmeans', KMeans(random_state=42))
])

In [ ]:
# Elbow method
K_values = list(range(1, 11))
wcss = []

for k in K_values:
    pipe.set_params(kmeans__n_clusters=k)
    pipe.fit(X_data)
    wcss.append(pipe['kmeans'].inertia_)

# Visualizing wcss versus k
fig = px.line(
    x=K_values,
    y=wcss,
    markers=True,
    title="Elbow Method Pipeline: StandardScaler + KMeans",
    labels={"x": "Number of Clusters (K)", "y": "Within-cluster Sum of Squares"}
)
fig.update_layout(height=500, width=700)
fig.show()

The plot of WCSS (within-cluster sum of squares) versus the number of clusters showed a steep decrease in WCSS from K=1 to K=3, followed by a slower, more gradual decline for higher K values. This “elbow” in the curve suggests that K=2 to K=5 clusters potentially capture most of the patterns in the data. Selecting K beyond 5 does not improve the grouping much and can overcomplicate the model.

In [10]:
# Silhouette scores to evaluate cluster quality
sil_scores = []
for k in range(2, 11):
    pipe.set_params(kmeans__n_clusters=k)
    pipe.fit(X_data)
    labels = pipe['kmeans'].labels_
    sil_scores.append(silhouette_score(X_data, labels))

fig = px.line(
    x=list(range(2, 11)),
    y=sil_scores,
    markers=True,
    title="Silhouette Scores for different K",
    labels={"x": "Number of Clusters (K)", "y": "Silhouette Score"}
)
fig.update_layout(height=500, width=700)
fig.show()

K's 3-10 have silhouette scores less than 0.25, which indicates weakly structured clusters that overlap. K=2 has a silhouette score of 0.88, which indicates clusters that are well-separated. 

In [11]:
# Fit the final k-means and cluster assignment
# Choosing a number of clusters based on elbow/silhouette
final_k = 2
pipe.set_params(kmeans__n_clusters=final_k)
pipe.fit(X_data)

# Assigning cluster labels to the dataframe
df['cluster'] = pipe['kmeans'].labels_

# Prints cluster sizes and scaled cluster centers
print("\nCluster sizes:")
print(df['cluster'].value_counts())

print("\nCluster centers scaled features:")
cluster_centers_df = pd.DataFrame(pipe['kmeans'].cluster_centers_, columns=X_cols)
print(cluster_centers_df)


Cluster sizes:
cluster
0    19512
1        2
Name: count, dtype: int64

Cluster centers scaled features:
   latitude  longitude  temperature_2m_mean  wind_speed_10m_mean  \
0  0.000085   0.000121            -0.000086             0.000083   
1 -0.832320  -1.182579             0.841281            -0.811702   

   precipitation_sum  fires_within_50km  fires_within_100km  \
0           0.000040          -0.008588           -0.005011   
1          -0.389057          83.780167           48.888815   

   distance_to_fire_km  fire_brightness  
0             0.000037         0.000051  
1            -0.362497        -0.496401  


Based on the silhouette scores, only 2 clusters were identified; however, the clustering is highly unbalanced. Cluster 0 contains 19,512 points, representing almost the entire dataset, while Cluster 1 contains only 2 points, indicating extreme outliers. 

Overall, these numbers indicate that K-Means with K=2 is not able to find well-separated, representative clusters for this dataset, and the algorithm is largely grouping the majority of the data together while isolating extreme cases.

This means that most of the data are within average conditions, and the K-means model is not able to find meaningful subgroups in the data or identifiable patterns.

In [12]:
# Average pm25 summary per cluster
print("\nAverage PM25 per cluster:")
print(df.groupby('cluster')['PM25'].mean())


Average PM25 per cluster:
cluster
0     7.418345
1    12.375000
Name: PM25, dtype: float64


The difference in PM2.5 between the clusters shows that the cluster 2 captures extreme pollution events, while cluster 1 captures normal air quality.

In [13]:
# Ridge/Lasso Regression

# Including cluster as a one-hot encoded feature and scaling features
X_supervised = pd.get_dummies(df[X_cols + ['cluster']], drop_first=True)
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_supervised)
y = df['PM25'].values

# Train/test/validation split
X_train, X_temp, y_train, y_temp = train_test_split(X_scaled, y, test_size=0.3, random_state=42)
X_test, X_val, y_test, y_val = train_test_split(X_temp, y_temp, test_size=0.5, random_state=42)

In [14]:
# Ridge regression with GridSearchCV
# Defining ridge and range of alpha values
ridge = Ridge()
ridge_params = {'alpha': np.logspace(-3, 3, 10)}

# GridSearchCV to find the best alpha using 5-fold cross-validation
ridge_grid = GridSearchCV(ridge, ridge_params, cv=5, scoring='r2')
ridge_grid.fit(X_train, y_train)
ridge_best = ridge_grid.best_estimator_

print("Best Ridge alpha:", ridge_grid.best_params_['alpha'])

Best Ridge alpha: 1000.0


Ridge regression shows that the model needs strong regularization to prevent overfitting.

In [15]:
# Lasso regression with GridSearchCV
# Defining lasso and range of alpha values
lasso = Lasso(max_iter=5000)
lasso_params = {'alpha': np.logspace(-3, 1, 10)}

# GridSearchCV to find the best alpha 
lasso_grid = GridSearchCV(lasso, lasso_params, cv=5, scoring='r2')
lasso_grid.fit(X_train, y_train)
lasso_best = lasso_grid.best_estimator_

print("Best Lasso alpha:", lasso_grid.best_params_['alpha'])

Best Lasso alpha: 0.05994842503189409


Lasso regression shows that the model needs mild regularization to prevent overfitting.

In [16]:
# Evaluate model performance through r-squared and rmse on train,test, and validation sets
def evaluate_model(model, X_train, y_train, X_test, y_test, X_val, y_val):
    results = {}
    for name, X_set, y_set in zip(['Train','Test','Validation'], 
                                  [X_train,X_test,X_val], 
                                  [y_train,y_test,y_val]):
        y_pred = model.predict(X_set)
        results[name] = {
            'R2': r2_score(y_set, y_pred),
            'RMSE': np.sqrt(mean_squared_error(y_set, y_pred))
        }
    return pd.DataFrame(results).T

print("\nRidge Performance:")
ridge_results = evaluate_model(ridge_best, X_train, y_train, X_test, y_test, X_val, y_val)
print(ridge_results)

print("\nLasso Performance:")
lasso_results = evaluate_model(lasso_best, X_train, y_train, X_test, y_test, X_val, y_val)
print(lasso_results)


Ridge Performance:
                  R2      RMSE
Train       0.127597  4.758092
Test        0.118096  5.064041
Validation  0.108358  4.759507

Lasso Performance:
                  R2      RMSE
Train       0.126474  4.761154
Test        0.115261  5.072174
Validation  0.106862  4.763498


Regularization helped slightly, but ultimately neither model captures a large portion of the variability in PM2.5. The models are robust but not highly predictive.

In [17]:
# Feature importance for ridge and alsso
ridge_coef = pd.Series(ridge_best.coef_, index=X_supervised.columns)
lasso_coef = pd.Series(lasso_best.coef_, index=X_supervised.columns)

print("\nRidge Coefficients:")
print(ridge_coef.sort_values(ascending=False))

print("\nLasso Coefficients (some may be zero due to regularization):")
print(lasso_coef.sort_values(ascending=False))


Ridge Coefficients:
temperature_2m_mean    0.406335
fires_within_100km     0.360417
longitude              0.159940
fires_within_50km      0.143734
fire_brightness       -0.115446
cluster               -0.255941
distance_to_fire_km   -0.443918
precipitation_sum     -0.449411
latitude              -0.630182
wind_speed_10m_mean   -1.004045
dtype: float64

Lasso Coefficients (some may be zero due to regularization):
temperature_2m_mean    0.379605
fires_within_100km     0.347658
longitude              0.106907
fires_within_50km      0.000000
fire_brightness       -0.073025
cluster               -0.099022
precipitation_sum     -0.420047
distance_to_fire_km   -0.440175
latitude              -0.620322
wind_speed_10m_mean   -1.026075
dtype: float64


Both models indicate that higher wind speed, higher latitude, greater precipitation, and longer distances from fires are associated with lower PM2.5. 

Higher temperature and more nearby fires (within 100 km) increase PM2.5. Ridge keeps all predictors, while Lasso removes less important ones, like fires within 50 km, highlighting the most influential variables. 

Overall, the models are robust and consistent across training, test, and validation sets, but they have limited predictive power, suggesting that other factors not included in the dataset may also affect PM2.5 levels.